crates/wasmtime/src/module.rs

```diff
@@ -535,6 +559,33 @@ impl Module {
     }
 }
 
+fn compiler_fingerprint(config: &Config) -> u64 {
+    use std::hash::Hasher;
+    let mut hasher = std::collections::hash_map::DefaultHasher::new();
+    config.compiler_fingerprint(&mut hasher);
+    hasher.finish()
+}
+
+/// Compile and write artifacts in I/O.
+pub fn compile_and_serialize(
```

Comment:
Would it be possible to add this as a method as `Module::serialize` instead of a top-level function?

Additionally, I think it may be best to stick to in-memory buffers for now since the entire `Module` has to be resident in memory anyway. Using `Read` and `Write` can be somewhat tricky so I'm not sure we want to go there just yet (we could always add support later though if needed)

Reply:
> Would it be possible to add this as a method as Module::serialize instead of a top-level function?

Currently, I'm discarding some artifacts once `CompiledModule` is created due to memory usage. Keeping them around maybe wasteful at the moment. Also it is why I chose `wasmtime_compile_and_serialize` instead of `wasmtime_module_serialize`.

Reply:
Could you expand a bit on what's not kept in memory? What's the scale of the memory usage improvement as well?

Reply:
> Could you expand a bit on what's not kept in memory?

Currently the answer is ELF image is modified by fixing relocation, so I'm just patching data directly. That makes image non-serializable as is. It needs to be re-cloned and "unpatched" to be serializable. Also there unwind_info we don't need to keep after `CompiledModule`. FWIW I'm currently working on both of these problems, but there are require some time and code to address.

Reply:
Hm ok from an API perspective I think it'd be best to have `Module::serialize` for now, and perhaps we can address the performance issues later?

Reply:
I took shortcut and start keeping data in `CompiledModule` that allows implement `Module::serialize`. I'll continue working on optimizing this part independent of this PR.